File: 0138. Copy List with Random Pointer/dfs.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* next;
    Node* random;
    
    Node(int _val) {
        val = _val;
        next = NULL;
        random = NULL;
    }
};
*/
class Solution {
public:
    // old Node - new Node
    map<Node*, Node*> visited;
    
    Node* copyRandomList(Node* head) {
        if(head == nullptr)
            return nullptr;
        if(visited.find(head) != visited.end())
            return visited[head];
        else{
            Node* root = new Node(head->val, nullptr, nullptr);
            visited[head] = root;
            
            root->next = copyRandomList(head->next);
            root->random = copyRandomList(head->random);
            return root;
        }
    }
};

File: 0138. Copy List with Random Pointer/dfs.java
/*
// Definition for a Node.
class Node {
    int val;
    Node next;
    Node random;

    public Node(int val) {
        this.val = val;
        this.next = null;
        this.random = null;
    }
}
*/

class Solution {
    private Map<Node, Node> mapping = new HashMap<>();
    public Node copyRandomList(Node head) {
        if(head == null)
            return null;
        if(mapping.containsKey(head)){
            return mapping.get(head);
        }else{
            int val = head.val;
            Node cur = new Node(val);
            mapping.put(head, cur);
            cur.next = copyRandomList(head.next);
            cur.random = copyRandomList(head.random);
            return cur;
        }
        
    }
}